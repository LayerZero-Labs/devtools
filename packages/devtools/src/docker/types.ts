export type ComposeSpecVersion = '3.9'

export interface ComposeSpecServiceHealthcheck {
    disable?: boolean
    interval?: string
    retries?: number
    test?: string | string[]
    timeout?: string
    start_period?: string
    start_interval?: string
}

export interface ComposeSpecServiceBuild {
    context?: string
    dockerfile?: string
    dockerfile_inline?: string
    args?: ListOrDict
    ssh?: ListOrDict
    labels?: ListOrDict
    cache_from?: string[]
    cache_to?: string[]
    no_cache?: boolean
    network?: string
    pull?: boolean
    target?: string
    privileged?: boolean
    tags?: string[]
    platforms?: string[]
}

export interface ComposeSpecPortDefinition {
    name?: string
    mode?: string
    host_ip?: string
    target?: number
    published?: string | number
    protocol?: string
}

export type ComposeSpecPortNumber = string | number

export type ComposeSpecPort =
    | `${ComposeSpecPortNumber}:${ComposeSpecPortNumber}`
    | ComposeSpecPortNumber
    | ComposeSpecPortDefinition

export interface ComposeSpecServiceVolumeDefinition {
    type: string
    source?: string
    target?: string
    read_only?: boolean
    consistency?: string
    bind?: {
        propagation?: string
        create_host_path?: boolean
        selinux?: 'z' | 'Z'
    }
    volume?: {
        nocopy?: boolean
    }
    tmpfs?: {
        size?: number | string
        mode?: number
    }
}

export type ComposeSpecServiceVolume = string | ComposeSpecServiceVolumeDefinition

export type ComposeSpecDependencyCondition = 'service_started' | 'service_healthy' | 'service_completed_successfully'

export interface ComposeSpecDependsOnDefinition {
    restart?: boolean
    required?: boolean
    condition: ComposeSpecDependencyCondition
}

export type ComposeSpecServiceDependsOn = string[] | Record<string, ComposeSpecDependsOnDefinition>

export type ComposeSpecServiceCommand = string | string[]

export interface ComposeSpecService {
    depends_on?: ComposeSpecServiceDependsOn
    build?: string | ComposeSpecServiceBuild
    image?: string
    expose?: ComposeSpecPortNumber[]
    command?: ComposeSpecServiceCommand
    ports?: ComposeSpecPort[]
    volumes?: ComposeSpecServiceVolume[]
    healthcheck?: ComposeSpecServiceHealthcheck
}

export type ComposeSpecServices = Record<string, ComposeSpecService>

export interface ComposeSpecVolumeDefinition {
    name?: string
    driver?: string
    driver_opts?: Record<string, string | number>
    labels?: ListOrDict
}

export type ComposeSpecVolume = null | ComposeSpecVolumeDefinition

export type ComposeSpecVolumes = Record<string, ComposeSpecVolume>

export interface ComposeSpec {
    version: ComposeSpecVersion
    services?: ComposeSpecServices
    volumes?: ComposeSpecVolumes
}

export type ListOrDict = Record<string, string | number | boolean | null> | string[]
