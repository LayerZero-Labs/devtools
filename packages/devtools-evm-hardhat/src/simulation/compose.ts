import { pipe } from 'fp-ts/lib/function'
import * as RR from 'fp-ts/ReadonlyRecord'
import type { ComposeSpec, ComposeSpecService } from '@layerzerolabs/devtools'
import { type AnvilOptions, createAnvilCliOptions } from '@layerzerolabs/devtools-evm'
import type { ComposeSpecServices } from '@layerzerolabs/devtools'
import type { SimulationConfig } from './types'

/**
 * Creates a docker compose service specification for an anvil-based EVM node
 *
 * @param {AnvilOptions} anvilOptions
 * @returns {ComposeSpecService}
 */
export const createEvmNodeServiceSpec = (anvilOptions: AnvilOptions): ComposeSpecService => ({
    // This service references a Dockerfile that is copied
    // next to the resulting docker-compose.yaml
    //
    // The source for this Dockerfile is located in src/simulation/assets/Dockerfile.conf
    build: {
        dockerfile: 'Dockerfile',
        target: 'node-evm',
    },
    command: ['anvil', ...createAnvilCliOptions(anvilOptions)],
})

/**
 * Creates a docker compose service specification for an nginx-based proxy service
 * that proxies requests to underlying EVM nodes (or their RPC URLs to be mor precise)
 *
 * @param {number} port
 * @param {ComposeSpecServices} networkServices
 * @returns {ComposeSpecService}
 */
export const createEvmNodeProxyServiceSpec = (
    port: number,
    networkServices: ComposeSpecServices
): ComposeSpecService => ({
    // This service references a Dockerfile that is copied
    // next to the resulting docker-compose.yaml
    //
    // The source for this Dockerfile is located in src/simulation/assets/Dockerfile.conf
    build: {
        dockerfile: 'Dockerfile',
        target: 'proxy-evm',
    },
    // This service will expose its internal 8545 port to a host port
    //
    // The internal 8545 port is hardcoded both here and in the nginx.conf file,
    // the source for which is located in src/simulation/assets/nginx.conf
    ports: [`${port}:8545`],
    depends_on: pipe(
        networkServices,
        // This service will depend on the RPCs to be healthy
        // so we'll take the networkServices object and replace
        // the values with service_healthy condition
        RR.map(() => ({
            condition: 'service_healthy',
        }))
    ),
})

/**
 * Creates a docker compose spec with a set of anvil-based EVM nodes
 * and a single proxy server that proxies requests to these nodes.
 *
 * @param {SimulationConfig} config
 * @param {Record<string, AnvilOptions>} networks
 * @returns {ComposeSpec}
 */
export const createSimulationComposeSpec = (
    config: SimulationConfig,
    networks: Record<string, AnvilOptions>
): ComposeSpec => ({
    version: '3.9',
    services: pipe(
        networks,
        // First we turn the networks into docker compose specs for EVM nodes
        RR.map(createEvmNodeServiceSpec),
        (networkServiceSpecs) =>
            // Then we add the RPC proxy server
            //
            // There is a small edge case here that we can address
            // if it ever comes up: if a network is called 'rpc', this compose file
            // will not work.
            //
            // The fix for this is to prefix all networks with something like network-xxx
            // but we can do that if ever this usecase comes up
            pipe(
                networkServiceSpecs,
                RR.upsertAt('rpc', createEvmNodeProxyServiceSpec(config.port, networkServiceSpecs))
            )
    ),
})
