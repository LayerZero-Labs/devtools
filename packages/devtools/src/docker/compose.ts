import { dump } from 'js-yaml'
import type { ComposeSpec } from './types'

export const serializeDockerComposeSpec = (spec: ComposeSpec): string => dump(spec)
