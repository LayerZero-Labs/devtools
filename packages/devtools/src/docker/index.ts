export * from './compose'
export * from './types'
