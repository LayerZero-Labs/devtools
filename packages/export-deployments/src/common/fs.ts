import { type Dirent, lstatSync, mkdirSync, opendirSync, readFileSync, writeFileSync } from 'fs'
import { resolve } from 'path'
import * as E from 'fp-ts/Either'
import { type Predicate } from 'fp-ts/lib/Predicate'
import { type OutputFile } from '../generator/types'

export const readFileSyncSafe = E.tryCatchK((path: string) => readFileSync(path, 'utf8'), E.toError)

export const isDirectorySafe = E.tryCatchK((path: string) => lstatSync(path).isDirectory(), E.toError)

export const mkdirSafe = E.tryCatchK((path: string): string => (mkdirSync(path, { recursive: true }), path), E.toError)

export const writeFileSyncSafe = E.tryCatchK(({ content, path }: OutputFile) => writeFileSync(path, content), E.toError)

/**
 * Lists all the directory entries that pass the predicate function. The returned array
 * will contain absolute paths of the directory entries.
 *
 * @param predicate `(dir: Dirent) => boolean`
 * @returns `(path: string) => Either<Error, string[]>`
 */
export const listEntriesSafe = (predicate: Predicate<Dirent>) =>
    E.tryCatchK((path: string) => {
        const dir = opendirSync(path)

        try {
            let subDirectory: Dirent | null
            const subDirectories: string[] = []

            while ((subDirectory = dir.readSync())) {
                if (predicate(subDirectory)) {
                    subDirectories.push(resolve(path, subDirectory.name))
                }
            }

            return subDirectories
        } finally {
            dir.closeSync()
        }
    }, E.toError)
