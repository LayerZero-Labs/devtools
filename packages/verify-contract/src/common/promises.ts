import assert from 'assert'

/**
 * Simple async retry functionality
 *
 * The onretry callback is async and will be awaited which allows for
 * custom timeout functionality
 */
export const retry = async <TResult>(
    callback: () => Promise<TResult>,
    numAttempts: number,
    onRetry?: (error: unknown, attempt: number) => unknown
): Promise<TResult> => {
    assert(numAttempts > 0, `Number of attempts for retry must be larger than 0, got ${numAttempts}`)

    // We'll try/catch the first N - 1 attempts, invoking onRetry after every failed attempt
    for (let attempt = 0; attempt < numAttempts - 1; attempt++) {
        try {
            return await callback()
        } catch (error) {
            await onRetry?.(error, attempt)
        }
    }

    // The last attempt we don't try/catch and let the error propagate to the user
    return callback()
}
