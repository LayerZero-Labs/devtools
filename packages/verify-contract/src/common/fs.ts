import { lstatSync } from 'fs'

export const isDirectory = (pathlike: string): boolean => {
    try {
        return lstatSync(pathlike).isDirectory()
    } catch {
        return false
    }
}

export const isFile = (pathlike: string): boolean => {
    try {
        return lstatSync(pathlike).isFile()
    } catch {
        return false
    }
}
