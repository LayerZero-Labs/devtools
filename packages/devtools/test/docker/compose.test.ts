import { serializeDockerComposeSpec } from '@/docker/compose'
import { spawnSync } from 'child_process'
import { rm, writeFile } from 'fs/promises'
import { join } from 'path'

describe('docker/compose', () => {
    describe('serializeDockerComposeSpec', () => {
        const SPEC_FILE_PATH = join(__dirname, 'docker-compose.yaml')

        const validateSpec = () => spawnSync('docker', ['compose', '-f', SPEC_FILE_PATH, 'config'])

        afterEach(async () => {
            await rm(SPEC_FILE_PATH, { force: true })
        })

        it('should create a valid compose spec if called with no services', async () => {
            const spec = serializeDockerComposeSpec({
                version: '3.9',
            })

            await writeFile(SPEC_FILE_PATH, spec)

            expect(validateSpec().status).toBe(0)

            expect(spec).toMatchSnapshot()
        })

        it('should create a valid compose spec if called with basic services', async () => {
            const spec = serializeDockerComposeSpec({
                version: '3.9',
                services: {
                    redis: {
                        image: 'redis:latest',
                    },
                    codebase: {
                        build: '.',
                        command: 'pnpm build',
                        volumes: ['./packages:/app/packages'],
                    },
                    postgres: {
                        image: 'postgres',
                        volumes: ['pgdata:/var/lib/postgresql/data'],
                    },
                },
                volumes: {
                    pgdata: null,
                    somevolume: {
                        name: 'somevolumename',
                    },
                },
            })

            await writeFile(SPEC_FILE_PATH, spec)

            expect(validateSpec().status).toBe(0)

            expect(spec).toMatchSnapshot()
        })
    })
})
