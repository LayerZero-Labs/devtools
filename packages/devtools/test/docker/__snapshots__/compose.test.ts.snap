// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`docker/compose serializeDockerComposeSpec should create a valid compose spec if called with basic services 1`] = `
"version: '3.9'
services:
  redis:
    image: redis:latest
  codebase:
    build: .
    command: pnpm build
    volumes:
      - ./packages:/app/packages
  postgres:
    image: postgres
    volumes:
      - pgdata:/var/lib/postgresql/data
volumes:
  pgdata: null
  somevolume:
    name: somevolumename
"
`;

exports[`docker/compose serializeDockerComposeSpec should create a valid compose spec if called with no services 1`] = `
"version: '3.9'
services: {}
"
`;
